lib/heapy/reference_explorer.rb

```diff
@@ -0,0 +1,161 @@
+require 'json'
+require 'readline'
+require 'set'
+
+module Heapy
+
+  # Follows references to given object addresses and prints
+  # them as a reference stack.
+  # Since multiple reference stacks are possible, it will preferably
+  # try to print a stack that leads to a root node, since reference chains
+  # leading to a root node will make an object non-collectible by GC.
+  #
+  # In case no chain to a root node can be found one possible stack is printed
+  # as a fallback.
+  class ReferenceExplorer
+    def initialize(filename)
+      @objects = {}
+      @reverse_references = {}
+      @virtual_root_address = 0
+      File.open(filename) do |f|
+        f.each.with_index do |line, i|
+          o = JSON.parse(line)
+          addr = add_object(o)
+          add_reverse_references(o, addr)
+          add_class_references(o, addr)
+        end
+      end
+    end
+
+    def drill_down_list(addresses)
+      addresses.each { |addr| drill_down(addr) }
+    end
+
+    def drill_down_interactive
+      while buf = Readline.readline("Enter address > ", true)
+        drill_down(buf)
+      end
+    end
+
+    def drill_down(addr_string)
+      addr = addr_string.to_i(16)
+      puts
+
+      chain = find_root_chain(addr)
+      unless chain
+        puts 'Could not find a reference chain leading to a root node. Searching for a non-specific chain now.'
+        puts
+        chain = find_any_chain(addr)
+      end
+
+      puts '## Reference chain'
+      chain.each do |ref|
+        puts format_object(ref)
+      end
+
+      puts
+      puts "## All references to #{addr_string}"
+      refs = @reverse_references[addr] || []
+      refs.each do |ref|
+        puts " * #{format_object(ref)}"
+      end
+
+      puts
+    end
+
+    def inspect
+      "<ReferenceExplorer #{@objects.size} objects; #{@reverse_references.size} back-refs>"
+    end
+
+    private
+
+    def add_object(o)
+      addr = o['address']&.to_i(16)
+      if !addr && o['type'] == 'ROOT'
+        addr = @virtual_root_address
+        o['name'] ||= o['root']
+        @virtual_root_address += 1
+      end
+
+      return unless addr
+
+      simple_object = o.select { |k, _v| %w[type file name class length imemo_type].include?(k) }
```

Comment:
Alternatively: @schneems how would you feel about dropping support for EOL rubies from `heapy`?

All CI failures so far were caused by tests for the older versions and this nit is the consequence of me dodging to use `slice`.

Also considering that `heapy` is not typically a dependency of a production application, but rather a tool running on a developer's machine I believe that it can allow itself to ask for recent Ruby versions.